stage0/src/main.rs

```diff
@@ -38,22 +38,32 @@ use x86_64::{
 };
 mod asm;
 
-// Magic constants where various data structures will be stored.
-// Many of these are set to be same as with crosvm; see
-// https://google.github.io/crosvm/appendix/memory_layout.html for more details.
-const BOOT_GDT_ADDR: PhysAddr = PhysAddr::new(0x1500);
-// The GDT provided by crosvm has 4 entries, which means that the (empty) IDT can go immediately
-// after that to address 0x1520. The Rust x86_64::GlobalDescriptorTable has 8 entries instead of 4,
-// and the x86_64::InterruptDescriptorTable takes a full 4K of memory, so let's move it to the next
-// page.
-const BOOT_IDT_ADDR: PhysAddr = PhysAddr::new(0x2000);
-const BOOT_PML4_ADDR: PhysAddr = PhysAddr::new(0x9000);
-const BOOT_PDPT_ADDR: PhysAddr = PhysAddr::new(0xA000);
-const BOOT_PD_ADDR: PhysAddr = PhysAddr::new(0xB000);
+#[link_section = ".boot.gdt"]
+static mut BOOT_GDT: MaybeUninit<GlobalDescriptorTable> = MaybeUninit::uninit();
+#[link_section = ".boot.idt"]
+static mut BOOT_IDT: MaybeUninit<InterruptDescriptorTable> = MaybeUninit::uninit();
+#[link_section = ".boot.pml4"]
+static mut BOOT_PML4: MaybeUninit<PageTable> = MaybeUninit::uninit();
+#[link_section = ".boot.pdpt"]
+static mut BOOT_PDPT: MaybeUninit<PageTable> = MaybeUninit::uninit();
+#[link_section = ".boot.pd"]
+static mut BOOT_PD: MaybeUninit<PageTable> = MaybeUninit::uninit();
+#[link_section = ".boot.unmeasured"]
+#[used]
```

Comment:
Does the used attribute actually do anything for these statics? I remember in your previous experiments it didn't seem to make a difference.

Reply:
Yes, they do! See the PR description for details, but this means that the Rust compiler won't throw them away. I think why it looked like they didn't do anything in my previous code was because the linker threw them away, but here we force both of them to include the seemingly unused data structures.